src/LtiAssignmentsGradesService.php

```diff
@@ -14,24 +14,48 @@ public function getScope(): array
         return $this->getServiceData()['scope'];
     }
 
+    // https://www.imsglobal.org/spec/lti-ags/v2p0#assignment-and-grade-service-claim
+    // When an LTI message is launching a resource associated to one and only one lineitem,
+    // the claim must include the endpoint URL for accessing the associated line item;
+    // in all other cases, this property must be either blank or not included in the claim.
+    public function getResourceLaunchLineItem(): ?LtiLineitem
+    {
+        $serviceData = $this->getServiceData();
+        if (empty($serviceData['lineitem'])) {
+            return null;
+        }
+
+        $lineitem = LtiLineitem::new()->setId($serviceData['lineitem']);
+        return $lineitem;
+    }
+
     public function putGrade(LtiGrade $grade, LtiLineitem $lineitem = null)
     {
         if (!in_array(LtiConstants::AGS_SCOPE_SCORE, $this->getScope())) {
             throw new LtiException('Missing required scope', 1);
         }
-        if ($lineitem !== null && empty($lineitem->getId())) {
-            $lineitem = $this->findOrCreateLineitem($lineitem);
-            $scoreUrl = $lineitem->getId();
-        } elseif ($lineitem === null && !empty($this->getServiceData()['lineitem'])) {
-            $scoreUrl = $this->getServiceData()['lineitem'];
-        } else {
-            $lineitem = LtiLineitem::new()
-                ->setLabel('default')
-                ->setScoreMaximum(100);
-            $lineitem = $this->findOrCreateLineitem($lineitem);
-            $scoreUrl = $lineitem->getId();
+
+        // If the line item already contains the Id, we do not need to find or create it.
+        if ($lineitem !== null) {
+            if (empty($lineitem->getId())) {
+                $lineitem = $this->findOrCreateLineitem($lineitem);
+            }
+        }
```

Comment:
```suggestion
        if (!isset($lineitem, $lineitem->getId()))
            $lineitem = $this->findOrCreateLineitem($lineitem);
        }
```
The `$lineitem->getId()` returns `null` if it's not set, so we're basically doing two checks null values here.

Reply:
I think this change will break the scenario when I pass in a specific $lineitem I want to use, that already has a URL in getId(). Assume $lineitem contains a line item that I have previous looked up with findLineItem() and then pass to putGrade(). Therefore both $lineitem and $lineitem->getId() are set, and this new code falls through to the else which will overwrite the $lineitem with the one from getResourceLaunchLineItem().
 
The intention with these separate checks is to handle the scenario where a non-null $lineitem is passed in with this if() statement, and use the scenario where no $lineitem is passed in in the else below this. But your change causes the else statement to trigger if $lineitem and $lineitem->getId() are both set, which is not correct.

Reply:
Ah. I think it should be `if (isset($lineitem) && !isset($lineitem->getId())`

Reply:
The catch is that in the case that lineitem is set (or not null), we want to leave it the same if it has an id, and find or create it, if it doesn't. But if it's set or not null, we don't want it falling through to the else statement where it would be set. Does that make sense? is there some clearer way in the code to express that? We can't combine these two into a single if statement or it changes the logic of when the else is entered (else not shown in this code snippet).